### PR TITLE
Installation time slots from candidate side

### PR DESCRIPTION
**Pull request by SaranchukVadym**

### CODE REVIEW

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+databaseChangeLog:
+  - changeSet:
+      id: 24.04.2021_add_time_slot_column_into_candidate_table
+      author: Saranchuk Vadym
+      preConditions:
+        onFail: MARK_RAN
+        columnExists:
+          tableName: CANDIDATE
+          columnName: CN_GRADUATION_DATE
```
Comment (ikantsevich) [inline sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml:9]: do we need check CN_GRADUATION_DATE column exists or CN_INTERVIEW_TIME?

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      changes:
+        - addColumn:
+            columns:
+              - column:
+                  name: CN_INTERVIEW_TIME
+                  afterColumn: CN_GRADUATION_DATE
+                  type: DATETIME
+                  defaultValue: NULL
```
Comment (ikantsevich) [inline sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml:17]: defaultValue null not need

File: sandbox-backend/src/main/resources/db/changelog/changes/07.05.2021_create_candidate_availability_time_table.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                  name: CAT_DATETIME
+                  type: DATETIME
+              - column:
+                  name: CAT_CREATED_AT
+                  type: DATETIME
+                  defaultValue: NOW()
+              - column:
+                  name: CAT_UPDATED_AT
+                  type: DATETIME
+                  defaultValue: NULL
```
Comment (ikantsevich) [inline sandbox-backend/src/main/resources/db/changelog/changes/07.05.2021_create_candidate_availability_time_table.yaml:35]: We need defaultValue: NOW() here

File: sandbox-backend/src/main/java/com/exadel/team2/sandbox/entity/CandidateEntity.java [hunk 4]
```diff
@@ -29,9 +28,6 @@
     @JoinColumn(name = "CN_ID", insertable = false, updatable = false)
     private InterviewTimeEntity interviewTimeEntity;
 
-//    @OneToMany(mappedBy = "candidateEntity", fetch = FetchType.EAGER)
-//    private Collection<EventEntity> eventEntities;
```
[the remark this answers is not in the document]
Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/entity/CandidateEntity.java:33]: I think it's not needed. Actually it doesn't work, I do sth wrong

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml (hunk 1 above)
Comment (SaranchukVadym) [reply sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml:9]: We should check CN_INTERVIEW_TIME column, it's my mistake

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml (hunk 2 above)
Comment (SaranchukVadym) [reply sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml:17]: What about data which already exists into this table? They won't have some value

File: sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        UpdateCandidateAvailabilityTime> implements CandidateAvailabilityTimeService {
+
+    private final CandidateDAO candidateDAO;
+    private final CandidateAvailabilityTimeDAO candidateAvailabilityTimeDAO;
+    private final ModelMap mapper;
+
+    @Override
+    public ResponseCandidateAvailabilityTimeDto getByCandidateId(Long id) {
+        List<CandidateAvailabilityTimeEntity> availabilityEntities =
+                ((CandidateAvailabilityTimeDAO)generalDAO).findAllByCandidateId(id);
```
[the remark this answers is not in the document]
Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java:36]: Yes, we can

[the remark this answers is not in the document]
Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java]: corrected

File: sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+        responseDto.setCandidateId(id);
+        availabilityEntities.forEach(dateTime -> responseDto.getAvailabilityTimeSlots().add(new TimeId(dateTime.getId(), dateTime.getDateTime())));
+
+        return null;
+    }
+
+    @Override
+    public void deleteAllByCandidateId(Long id) {
+        ((CandidateAvailabilityTimeDAO)generalDAO).deleteAllByCandidateId(id);
```
[the remark this answers is not in the document]
Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java]: corrected

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml (hunk 1 above)
Comment (SaranchukVadym) [reply sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml:9]: corrected

File: sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml (hunk 2 above)
Comment (SaranchukVadym) [reply sandbox-backend/src/main/resources/db/changelog/changes/06.05.2021_add_time_slot_column_into_candidate_table.yaml]: corrected

File: sandbox-backend/src/main/resources/db/changelog/changes/07.05.2021_create_candidate_availability_time_table.yaml (hunk 3 above)
Comment (SaranchukVadym) [reply sandbox-backend/src/main/resources/db/changelog/changes/07.05.2021_create_candidate_availability_time_table.yaml]: corrected

File: sandbox-backend/src/main/java/com/exadel/team2/sandbox/entity/CandidateEntity.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/entity/CandidateEntity.java:33]: corrected

File: sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        ResponseCandidateAvailabilityTimeDto responseDto = new ResponseCandidateAvailabilityTimeDto();
+
+        if (availabilityEntities == null) {
+            return responseDto;
+        }
+
+        responseDto.setCandidateId(id);
+        availabilityEntities.forEach(dateTime -> responseDto.getAvailabilityTimeSlots().add(new TimeId(dateTime.getId(), dateTime.getDateTime())));
+
+        return null;
```
Comment (ikantsevich) [inline sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java:46]: Why null, but not responseDTO?

Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java]: It my mistake

Comment (SaranchukVadym) [reply sandbox-backend/src/main/java/com/exadel/team2/sandbox/service/impl/CandidateAvailabilityTimeServiceImpl.java]: corrected

**Review by ikantsevich — APPROVED** [review]

**Pull request merged**